### DISCUSSION TITLE
defining multiple blocks (block IDs) based on an image file

### DISCUSSION THREAD

**Discussion by Ali-toghraee** [idaholab/moose#17298] [Q&A Meshing] [answered]
Dear moose users,

 I am trying to create a semi poly-crystalline (multiphase) domain. Such that each I can assign each color a block id and that later I can use those IDs for my initial conditions. Some thing like the pic below.
![mesh00](https://user-images.githubusercontent.com/77089229/110874057-49345500-8290-11eb-842f-7838b9559102.png)

I am trying to use the "ImageSubdomainGenerator" . But I have a problem in assigning IDs to different colors and I only get 2 ids (high and low) as below

![mesh_out](https://user-images.githubusercontent.com/77089229/110874442-fa3aef80-8290-11eb-8e12-14228b8f06f8.png)

Could any body help me how to do that. I have also included some part of the code below:

[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 200
    ny = 100

    xmax = 200
    xmin = 0

    ymax = 100
    ymin = 0

    elem_type = QUAD4
  []
  [./im1]
    input = gen
    type = ImageSubdomainGenerator
    file = mesh00.png
    threshold = 90
    lower_value = 6
    upper_value = 7

  [../]

  [./im2]
    input = im1
    type = ImageSubdomainGenerator
     file = mesh00.png
    threshold = 70
    lower_value = 4
     upper_value = 5

  [../]

  [./im3]
    input = im2
    type = ImageSubdomainGenerator
     file = mesh00.png
    threshold = 25
    lower_value = 2
     upper_value = 3

  [../]
[]

[Variables]
  [u]
  []
[]

[Problem]
  type = FEProblem
  solve = false
[]

[Executioner]
  type = Steady
[]

[Outputs]
  execute_on = 'timestep_end'
  exodus = true
[]

**Comment by Ali-toghraee** [comment]
@WilkAndy

**Reply by WilkAndy** [reply]
Hey @Ali-toghraee , i'm not really an expert in this.  did you tag me for a particular reason?

**Reply by Ali-toghraee** [reply]
Hi Andy, 

Sorry to tag you. Since you helped me last time on "modeling a multi phase multi-shape initial condition" , I thought you could help me on this. Do you know who can help on this matter? 

Best

**Reply by WilkAndy** [reply]
Let's just see who answers....

**Comment by ngrilli** [comment, marked as answer]
Dear @Ali-toghraee 

I believe your ImageSubdomainGenerator blocks are cancelling each other, therefore you
only see the effect of the last one: so block IDs are only 2 and 3.

If you check the code MeshBaseImageSampler,
you will see that "upper_value" is the value set if the corresponding grayscale is higher than "threshold"
while "lower_value" is the value set if the grayscale color is lower than "threshold".

The way you define your blocks is such that the next block always cancel the previous one
because only two block IDs are assigned to the entire geometry each time.

You could try to restrict the ImageSubdomainGenerator objects to one of the block that is created by the
previous ImageSubdomainGenerator object and see if that works.

I mean you can try to add 
block = 7 (or block = 6)
to [./im2]
block = 5 (or block = 4)
to [./im3]
et cetera

Best Regards,
Nicolò Grilli
National University of Singapore

**Reply by Ali-toghraee** [reply]
Hello @ngrilli 

Your are exactly right about my ImageSubdomainGenerator blocks  cancelling each other. But I am not sure how to restrict them.
Because ,as you mentioned, each time they assign only 2 values ( high and low). I think we need to  define the input for the [./im2] be  " block 6"  instead of the whole [./im1] , but I dont know how to do that.
Could you explain more on how to add "block = x" inside the  ImageSubdomainGenerator?
when I add "block = x" , I see the error:

*** ERROR ***
/home/atoghraee/projects02/ox_poly/poly/mesh.i:40: unused parameter 'Mesh/im2/block'
/home/atoghraee/projects02/ox_poly/poly/mesh.i:52: unused parameter 'Mesh/im3/block'

**Reply by Ali-toghraee** [reply]
Update:

I changed my input file to:

[Mesh]
  [gen]

    type = GeneratedMeshGenerator
    dim = 2
    nx = 200
    ny = 100

    xmax = 200
    xmin = 0

    ymax = 100
    ymin = 0

    elem_type = QUAD4
  []
  [./im1]
    input = gen
    type = ImageSubdomainGenerator
    file = mesh_gray.png
     # threshold = 20
     lower_value = 1
     upper_value = 6
  [../]
  #
  # [./im2]
  #   input = im1
  #   type = ImageSubdomainGenerator
  #     file = mesh_gray.png
  #   # threshold = 50
  #   # block = 6
  #   # lower_value = 4
  #   # upper_value = 5
  # [../]
  #
  # [./im3]
  #   input = im2
  #   type = ImageSubdomainGenerator
  #     file = mesh_gray.png
  #   #  threshold = 25
  #   #  lower_value = 2
  #   #   upper_value = 3
  #   # block = 5
  # [../]
[]

[Variables]
  [u]
  []
[]

[Problem]
  type = FEProblem
  solve = false
[]

[Executioner]
  type = Steady
[]

[Outputs]
  execute_on = 'timestep_end'
  exodus = true
[]

now for the input picture of:
![mesh_gray](https://user-images.githubusercontent.com/77089229/110994646-4b50ef00-8336-11eb-9b10-71dc9708a51d.png)

I get the result of:
![mesh_gray_result](https://user-images.githubusercontent.com/77089229/110994884-bc90a200-8336-11eb-9c64-5a84ddfa14a3.png)

- The question is how to map 0-225 to the 1-6 block numbers
- And why the numbering is not consistent  with the gray scale! (darker=> lower , lighter => higher or vise versa)

**Reply by ngrilli** [reply]
Dear @Ali-toghraee 

Apologies, I did not know that "block" cannot be used in ImageSubdomainGenerator.

In the base class MeshBaseImageSampler, there is the option "component", which means the RGB component.
If this is not set, the greyscale value is used, but I don't know how RGB is converted to greyscale.

I suggest you try setting "component = 0" inside your ImageSubdomainGenerator blocks,
so the red colour will be read. Then you can assign increasing values of the red colour to your different domains
by modifying the image. 

Best Regards,
Nicolò Grilli
National University of Singapore
